materials_io/utils/interface.py

```diff
@@ -94,21 +124,55 @@ def get_adapter(name: str) -> BaseAdapter:
     return mgr.driver
 
 
-def execute_parser(name, group, context=None, adapter=None):
-    """Invoke a parser on a certain group of data
+def execute_parser(name, group, global_context=None,
+                   adapter_context: Union[None, dict] = None,
+                   parser_context: Union[None, dict] = None,
+                   adapter: Union[None, str] = None,
+                   adapter_map: Union[None, str] = None,
+                   default_adapter: Union[None, str] = None):
+    """Invoke a parser on a certain group of data AND apply appropriate adapter to metadata.
 
     Args:
         name (str): Name of the parser
         group ([str]): Paths to group of files to be parsed
-        context (dict): Context of the files, used in adapter and parser
+        global_context (dict): Context of the files, used for every parser and adapter
+        adapter_context (dict): Context used for adapters. Key is the name of the adapter,
+            value is the context.  The key ``@all`` is used to for context used for every adapter
+        parser_context (dict): Context used for adapters. Key is the name of the parser,
+            value is the context. The key ``@all`` is used to for context used for every parser
         adapter (str): Name of adapter to use to transform metadata
+        adapter_map (dict): keys are adapter/parser names, values are adapters
+        default_adapter (str): Adapter to use if no other adapter is defined
     Returns:
         ([dict]): Metadata generated by the parser
     """
-    metadata = get_parser(name).parse(group, context)
+
+    # Load in default arguments
+    if global_context is None:
+        global_context = dict()
+    if adapter_context is None:
+        adapter_context = dict()
+    if parser_context is None:
+        parser_context = dict()
+
+    my_parser_context, my_adapter_context = _get_adapter_parser_context(global_context,
+                                                                        parser_context,
+                                                                        adapter_context,
+                                                                        name)
+
+    metadata = get_parser(name).parse(group, my_parser_context)
     if adapter is not None:
-        adapter = get_adapter(adapter)
-        return adapter.transform(metadata, context=context)
+        parser = get_parser(name)
+        adapter_map = get_adapter_map(adapter_map=adapter_map, parsers=[parser])
+
+        adapter_name = adapter_map.get(name, default_adapter)
+        if adapter_name is not None:
+            adapter = get_adapter(adapter_name)
+            try:
+                metadata = adapter.transform(metadata, my_adapter_context)
+            except Exception:
```

Comment:
It is OK for this function to throw an Exception. There is no contract that the function returns `None` if the parsers fail to work.

Reply:
Plus, line 163 will through an exception

Reply:
Rolled back this function entirely, so should allow exceptions to propagate to the top of the stack.